### PR TITLE
Enable whisper encoder to accept any chunk length

### PR DESCRIPTION
**Pull request by MahmoudAshraf97**
# What does this PR do?

Enables Whisper model and encoder to run using any chunk length <= 30s, as for now, the only method to change the chunk length is to use a pipeline which might not be suitable if running a custom generation loop is needed ( continuous batching for example)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@sanchit-gandhi

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @kamilakesbi

**Comment by MahmoudAshraf97** [comment]
As for torch compile, the behavior before and after is identical and I tested that,

**Comment by MahmoudAshraf97** [comment]
@kamilakesbi 
1. As for torch compile, the behavior before and after is identical and I tested that
2. the effect of shorter chunks on whisper performance is thoroughly studied in [Distil-Whisper-Large-V2](https://huggingface.co/distil-whisper/distil-large-v2) and [Distil-Whisper-Large-V3](https://huggingface.co/distil-whisper/distil-large-v3) model cards and the paper, which use 15s and 25s respectively instead of 30s

The chunk size can already be changed when using whisper through a pipeline, and the way that works is that the features are padded with zeros  to reach 30s and this is very inefficient given that there's no need for that if the encoder can accept inputs less than 30s which is tested in other frameworks such as `tensorrt_llm` and `faster-whisper`

Merging this PR will enable us to use a pipeline without padding, speeding up the inference of the encoder part and using less memory, and also use the model directly without needing to use a pipeline just to change the chunk length

**Comment by kamilakesbi** [comment]
@MahmoudAshraf97 thanks for explaining the benefits of using shorter chunk length! :) 

It seems you've already studied how adding this feature would affect compatibility with `torch.compile` and Whisper's performance! 

Could you please add a test to make sure the model gives the same output for audios <30 sec with and without padding (if you could add a test with batched audios that would be great). 

cc @ylacombe I would like your opinion on this feature request / PR

**Comment by MahmoudAshraf97** [comment]
Please note that this PR doesn't remove padding, it just gives the ability to do so without having to use a pipeline. 
I experimented with removing the padding from the pipeline but it wasn't as trivial as I expected, as the chunking algorithm used just divides the audio with overlapping strides and that is known to cause hallucinations with whisper, when a vad chunking algoritm is used, the output is almost identical to the padded input.
I tested that by chunking the audio without relying on the current chunking algorithm, then using the feature extractor to generate the input mels, then use `.generate` method and the results were good with speedups gained from the faster encoder inference.
My experiments with removing pipeline padding included evaluating `librispeech_asr` validation set with 15s, 25s and 30s chunk lengths
15s: 8.7 WER
25s: 3.61 WER
30s: 3.61 WER

in this particular test case, 15s performed much worse because most audios are slightly longer than 15s so the second window performs badly, with longer files (>60s), I experienced no performance degradation

TLDR:  this PR is backwards compatible, no behavior changes are expected for all usecases except when using `WhisperModel.encoder.forward`  or `WhisperModel.generate` directly which isn't a popular use case, but I need it for a tutorial that I'm making that explains how to implement continuous batching using HF transformers (3x speedup!, advice is appreciated on where it should be placed in the examples folder)

**Comment by ylacombe** [comment]
Hi @MahmoudAshraf97, thanks for opening this PR!

I'm not sure to follow where you removed padding in the code you've shared for benchmark tbh, could you elaborate?
Also, using the pipeline here adds an additional complexity to what you're doing. Could you possibly do a benchmark on audio degradation using the processor and the model (without using the pipeline)?
 

Note that I'm not against making Whisper accept any chunk length, as long as we thoroughly verify that it's backward compatible, that the usage is justified and that the original behavior is enabled by default! Let's make sure we verify all previous conditions

**Comment by MahmoudAshraf97** [comment]
Sorry for the confusion, this PR doesn't remove padding, to do so, add `max_length=chunk_len` to this feature extractor call
https://github.com/huggingface/transformers/blob/01be5b48790f113b7d71943b580c842e3e097988/src/transformers/pipelines/automatic_speech_recognition.py#L66

The use of pipeline was intended because it was the only way to change chunk length available currently in `transformers`
As for the other concerns

Backward compatibility is confirmed because it still allows the old behavior and the test suite pass 100%, as for whether it's justified or not, IMO I see no reason for that restriction in the first place as it doesn't exist in either the original whisper implementation neither other custom implementations, so it makes using the whisper model directly without a pipeline more constrained, and as you said, using a pipeline adds an additional complexity

**Comment by sanchit-gandhi** [comment]
Hey @MahmoudAshraf97! Thanks for opening this PR and for the justifications on why it should be added. Note that we're mixing two definitions here, which is potentially causing confusion:
1. **Context window:** this refers to the receptive field of the Whisper encoder, i.e. the length of the audios it can process at once. In both the original and Transformers implementations, this is fixed to 30-seconds. Any audios less than 30-seconds are padded to 30-seconds and passed through the encoder. Any audios longer than 30-seconds need to be truncated, or passed sequentially through the encoder for sequential long-form generation. Why do we this at inference time? Simply because this is how it's done at training time: audios are padded/truncated as described above, so the model learns to operate with a 30-second context window. While it is possible to reduce the context window dynamically during inference time, e.g. by slicing the encoder positional embeddings to the length of the audio inputs, this leads to a degradation in performance compared to a fixed 30-second context window. You can read the results in [this repo](https://github.com/futo-org/whisper-acft#results) for details.
2. **Chunk length:** this refers to our strategy for dividing up a long audio file into smaller segments, and passing the segments in parallel through the model. Each segment is still padded/truncated to the model's context window before being passed through the encoder.

To understand how to set the context window and chunk length, we can take the examples of the Whisper model and Distil-Whisper models:
1. Whisper: the model was trained with a fixed 30-second context window, and the average audio length during training was likely close to 30-seconds. This means at inference time, we use a fixed context window of 30-seconds and a chunk length of 30-seconds.
2. Distil-Whisper large-v2: trained with a fixed 30-second context window (same as Whisper), but the average audio length was 7.1-seconds (see Table 13 of the [paper](https://arxiv.org/pdf/2311.00430). At inference time, we use a fixed context window of 30-seconds, but a chunk length of 15-seconds (an interpolation between the 30-second audios used to pre-train Whisper, and the 7-second audios used to distill it)[^1]
3. Distil-Whisper large-v3: trained with a fixed 30-second context window (same as Whisper), but the average audio length during training was much closer to 30-seconds. At inference time, we use a fixed context window of 30-seconds, but a chunk length of 25-seconds.

=> you can see that in all cases, the models were trained on fixed 30-second context windows, so the best performance is obtained by using these fixed window sizes. However, the chunk length is dependent on the distribution of audio data used during training.

In relation to this PR: we're changing the context window, but not the chunk length (which is controlled by the pipeline parameter `chunk_length_s`). Given this typically leads to a performance degradation using the pre-trained models, I would be in favour of not adding it to the main codebase, but you're free to experiment with these changes in a fork and benchmark performance!

Related:
1. #25744
2. #26214

----
[^1]You're probably wondering: why did you use a 30-second context window if your audios were on average 7.1-seconds in duration? We tried distilling using a smaller 15-second context window, but similar to dynamically changing the window, this hurt performance. The best performance was always obtained by copying the Whisper encoder exactly from the original OpenAI implementation.

**Comment by MahmoudAshraf97** [comment]
Thanks for the insightful comment @sanchit-gandhi , I just verified by evaluating `distil-large-v2` with `chunk_length=15` on [this dataset](https://huggingface.co/datasets/mobiuslabsgmbh/youtube-commons-asr-eval) which is mostly longform audio.
With padding: WER: 13.433
Without padding: WER: 19.435
